Make force/energy kernels pure functions

jngrad opened the pull request:
Fixes #2502, fixes #3033

Description of changes:
- make force/energy kernels pure functions
- use `boost::optional` for breakable bonded IAs

jngrad committed:
Remove output parameters in force/energy kernels

Force/energy kernels now return forces and torques as tuples of
Vector3d instead of altering the particles passed as arguments.

jngrad committed:
Fix regression

jngrad committed:
Fix regression (error: jump to case label)

jngrad committed:
Use boost::optional for breakable bonded IAs

jngrad committed:
Simplify signature of angle force kernels

Force kernels `calc_angle_*_3body_forces()` now take particle
positions instead or particle pointers as arguments.

jngrad committed:
Merge branch 'python' into fix-2502-p2

jngrad committed:
Update doxygen documentation

jngrad committed:
Update doxygen documentation

jngrad committed:
More refactoring

Use vector_product(), bool type for function flags, and
const Particle in energy/force kernels that don't modify particles.

jngrad commented:
I'm opening this as WIP because reviews might find ways to improve the code further.

OIF global forces and some ELC kernels aren't pure functions.

Here's the LJ performance:

|  commit |            label | empty | default | maxset |
| ------- | ---------------- | ----- | ------- | ------ |
| 2d5e097 |  boost::optional |  65%  |  108%   | 114%   |
| 7e10ad0 |           python |  61%  |  100%   | 105%   |

I'm not displaying the P3M results because the non-deterministic nature of the tuning function is responsible for the large performance fluctuations (up to 27 percentage points) observed in previous benchmarks.

jngrad committed:
Remove legacy debug macros in bonded/non-bonded IAs

jngrad committed:
Fix regression in 53b8741888

KaiSzuttor commented:
If we'd really wanted to improve the performance we would probably need to use some kind of expression templates?

RudolfWeeber commented:
On Fri, Aug 23, 2019 at 03:29:29AM -0700, Kai Szuttor wrote:
> If we'd really wanted to improve the performance we would probably need to use some kind of expression templates?
We should probably also figure out, whether it is really necessary to be able to use more than one short-rnage kernel for a given pair of particle types.
I personally never did that and I'm not aware of a resonable application.

jngrad committed:
Merge branch 'fix-3034' into fix-2502-p2

jngrad committed:
Refactor rotation

jngrad committed:
Style patch, parameter S

jngrad committed:
Merge branch 'python' into fix-2502-p2

jngrad committed:
Rename non-bonded force kernels

jngrad committed:
More efficient summation of non-bonded forces

Reduce a list of doubles instead of reducing a list of intermediate
Vector3d objects in calc_non_bonded_pair_force_parts().

jngrad commented:
I don't see how to improve performance further in this PR. The commit 3ea4ac6 managed to soften the performance impact of making the kernels pure functions, however it will need to be reverted in espresso 5: once we refactor the non-bonded IAs logic using switches or variants to only evaluate those kernels that are active, this summation will go away.

|  commit |   label | empty | default |  maxset |
| ------- | ------- | ----- | ------- | ------- |
| 3ea4ac6 |  reduce |  67%  |   105%  |   113%  |
| 6374bc7 |  merge  |  68%  |   109%  |   115%  |
| e2e7c5d |  python |  63%  |   100%  |   105%  |

edit: revert 3ea4ac6 + 4d71264 in espresso 5

jngrad committed:
Clang-tidy

KaiSzuttor commented on src/core/bonded_interactions/bonded_coulomb.hpp:
```diff
@@ -44,34 +44,30 @@ int bonded_coulomb_set_params(int bond_type, double prefactor);
-    double const q1q2, Bonded_ia_parameters const *const iaparams,
-    Utils::Vector3d const &dx, Utils::Vector3d &force) {
+inline boost::optional<Utils::Vector3d>
+calc_bonded_coulomb_pair_force(double const q1q2,
```
why is there a ```calc``` before the force calculating function's names but not for the energy kernels? Maybe we should drop the prefix.

jngrad replied:
I've now removed the prefix in force/energy kernels

KaiSzuttor commented on src/core/bonded_interactions/bonded_coulomb.hpp:
```diff
@@ -44,34 +44,30 @@ int bonded_coulomb_set_params(int bond_type, double prefactor);
-    Utils::Vector3d const &dx, Utils::Vector3d &force) {
+inline boost::optional<Utils::Vector3d>
+calc_bonded_coulomb_pair_force(double const q1q2,
+                               Bonded_ia_parameters const *const iaparams,
```
can't we pass the parameters by ```const &```?

jngrad replied:
Yes, and this made the integration loop a lot more homogeneous. Only the cell system and the principal loop of files `*_inline.hpp` use Particle pointers now, because in these parts of the code we don't always know in advance if a particle exists.

KaiSzuttor commented on src/core/bonded_interactions/bonded_coulomb_sr.hpp:
```diff
@@ -46,39 +46,30 @@ int bonded_coulomb_sr_set_params(int bond_type, double q1q2);
-                              Bonded_ia_parameters const *const iaparams,
-                              Utils::Vector3d const &dx, double *_energy) {
+inline boost::optional<double> bonded_coulomb_sr_pair_energy(
+    Particle const *const p1, Particle const *const p2,
```
can we pass the particles by ```const &```?

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -89,19 +87,18 @@ inline bool calc_tab_bond_force(Bonded_ia_parameters const *const iaparams,
   if (dist < tab_pot->cutoff()) {
-    *_energy = tab_pot->energy(dist);
-    return false;
+    auto const energy = tab_pot->energy(dist);
```
just return here

jngrad replied:
Done

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -64,21 +64,19 @@ int tabulated_bonded_set_params(int bond_type,
     auto const fac = tab_pot->force(dist) / dist;
-    force = fac * dx;
-    return false;
+    auto const force = fac * dx;
```
just return here, the variable ```force``` is not needed

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -176,8 +166,8 @@ inline bool tab_angle_energy(Utils::Vector3d const &r_mid,
 #endif
-  *_energy = iaparams->p.tab.pot->energy(phi);
-  return false;
+  auto const energy = iaparams->p.tab.pot->energy(phi);
```
return here

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -263,9 +242,8 @@ inline bool tab_dihedral_energy(Utils::Vector3d const &r1,
-  *_energy = tab_pot->energy(phi);
-
-  return false;
+  auto const energy = tab_pot->energy(phi);
```
return here

KaiSzuttor commented on src/core/bonded_interactions/dihedral.hpp:
```diff
@@ -197,12 +188,16 @@ dihedral_energy(Utils::Vector3d const &r1, Utils::Vector3d const &r2,
+  }
 
-  *_energy =
+  auto const energy =
```
return here

KaiSzuttor commented on src/core/bonded_interactions/fene.hpp:
```diff
@@ -59,31 +57,29 @@ inline bool calc_fene_pair_force(Bonded_ia_parameters const *const iaparams,
   }
 
-  force = fac * dx;
+  auto const force = fac * dx;
```
return here

KaiSzuttor commented on src/core/bonded_interactions/fene.hpp:
```diff
@@ -59,31 +57,29 @@ inline bool calc_fene_pair_force(Bonded_ia_parameters const *const iaparams,
-  energy *= log((1.0 - dr * dr * iaparams->p.fene.drmax2i));
-  *_energy = energy;
-  return false;
+  auto const energy = -0.5 * iaparams->p.fene.k * iaparams->p.fene.drmax2 *
```
return here

KaiSzuttor commented on src/core/bonded_interactions/harmonic.hpp:
```diff
@@ -60,29 +58,28 @@ inline bool calc_harmonic_pair_force(Bonded_ia_parameters const *const iaparams,
     fac = 0;
   }
-  force = fac * dx;
+  auto const force = fac * dx;
```
return here

KaiSzuttor commented on src/core/bonded_interactions/harmonic.hpp:
```diff
@@ -60,29 +58,28 @@ inline bool calc_harmonic_pair_force(Bonded_ia_parameters const *const iaparams,
   }
 
-  *_energy =
+  auto const energy =
```
return here

KaiSzuttor commented on src/core/bonded_interactions/harmonic_dumbbell.hpp:
```diff
@@ -46,60 +46,56 @@ int harmonic_dumbbell_set_params(int bond_type, double k1, double k2, double r,
-                 Utils::sqr(dist - iaparams->p.harmonic.r) +
-             0.5 * iaparams->p.harmonic_dumbbell.k2 * (torque * diff);
-  return false;
+  auto const energy = 0.5 * iaparams->p.harmonic_dumbbell.k1 *
```
return here

KaiSzuttor commented on src/core/bonded_interactions/quartic.hpp:
```diff
@@ -42,46 +42,42 @@ int quartic_set_params(int bond_type, double k0, double k1, double r,
-  force = -fac * dx;
-
-  return false;
+  auto const force = -fac * dx;
```
return here

KaiSzuttor commented on src/core/bonded_interactions/quartic.hpp:
```diff
@@ -42,46 +42,42 @@ int quartic_set_params(int bond_type, double k0, double k1, double r,
-  *_energy = 0.5 * iaparams->p.quartic.k0 * dr2 +
-             0.25 * iaparams->p.quartic.k1 * Utils::sqr(dr2);
-  return false;
+  auto const energy = 0.5 * iaparams->p.quartic.k0 * dr2 +
```
return here

KaiSzuttor commented on src/core/bonded_interactions/subt_lj.hpp:
```diff
@@ -43,28 +43,23 @@ int subt_lj_set_params(int bond_type);
+inline boost::optional<Utils::Vector3d>
+calc_subt_lj_pair_force(IA_parameters const *const iaparams,
+                        Utils::Vector3d const &dx) {
+  auto const force = -lj_pair_force(iaparams, dx, dx.norm());
```
return here

KaiSzuttor commented on src/core/bonded_interactions/subt_lj.hpp:
```diff
@@ -43,28 +43,23 @@ int subt_lj_set_params(int bond_type);
+inline boost::optional<double>
+subt_lj_pair_energy(IA_parameters const *const iaparams,
+                    Utils::Vector3d const &dx) {
+  auto const energy = -lj_pair_energy(iaparams, dx.norm());
```
return here

KaiSzuttor commented on src/core/bonded_interactions/umbrella.hpp:
```diff
@@ -48,31 +48,29 @@ inline double umbrella_force_r(double k, int dir, double r, double distn) {
-  *_energy = 0.5 * ia_params->p.umbrella.k *
-             Utils::sqr(distn - ia_params->p.umbrella.r);
-  return false;
+  auto const energy = 0.5 * ia_params->p.umbrella.k *
```
return here

KaiSzuttor commented on src/core/electrostatics_magnetostatics/coulomb_inline.hpp:
```diff
@@ -52,30 +52,34 @@ inline Utils::Vector3d central_force(double const q1q2,
+  }
 
-  force += central_force(q1q2, d, dist);
+  Utils::Vector3d force = central_force(q1q2, d, dist);
```
```auto const force```

mkuron commented on src/core/electrostatics_magnetostatics/p3m-dipolar.hpp:
```diff
@@ -221,11 +221,12 @@ void dp3m_shrink_wrap_dipole_grid(int n_dipoles);
-                                  double dist, Utils::Vector3d &force) {
+inline std::tuple<double, Utils::Vector3d, Utils::Vector3d, Utils::Vector3d>
+dp3m_add_pair_force(Particle const *const p1, Particle const *const p2,
+                    Utils::Vector3d const &d, double dist2, double dist) {
```
Function needs to be renamed from `add` to `calc`.

KaiSzuttor replied:
i'd suggest to drop both: ```dp3m_pair_force```

jngrad replied:
Thanks! I've now removed the `add` part.

KaiSzuttor commented on src/core/electrostatics_magnetostatics/p3m-dipolar.hpp:
```diff
@@ -268,19 +270,23 @@ inline double dp3m_add_pair_force(Particle *p1, Particle *p2,
+    return std::make_tuple(_energy, force, torque1, torque2);
   }
-  return 0.0;
+  return std::make_tuple(0.0, Utils::Vector3d{}, Utils::Vector3d{},
```
why is there sometimes an energy returned and sometimes not and why is it ever returned if it's not used anywhere in the code?

jngrad replied:
I don't know why NPT requires energies during the calculation of forces. More worrisome here is the code duplication of the energy calculation in `dp3m_add_pair_force()` and `dp3m_add_pair_energy()`, the latter containing comments ([1](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/p3m-dipolar.hpp#L311), [2](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/p3m-dipolar.hpp#L336-L337)) suggesting the energy formula used in the former ([3](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/p3m-dipolar.hpp#L280-L282)) is wrong.

mkuron commented on src/core/electrostatics_magnetostatics/dipole_inline.hpp:
```diff
@@ -8,33 +8,36 @@
+inline double add_pair_energy(Particle const *const p1,
+                              Particle const *const p2,
+                              Utils::Vector3d const &d, double dist,
+                              double dist2) {
```
Again, this function now calculates instead of adding.

jngrad replied:
I've now removed the prefix.

KaiSzuttor commented on src/core/electrostatics_magnetostatics/dipole_inline.hpp:
```diff
@@ -8,33 +8,36 @@
     // fall trough
   case DIPOLAR_P3M: {
+    double eng;
+    std::tie(eng, force, torque1, torque2) =
```
if there was not this energy part in the returned tuple we could just do ```return dp3m_add_pair_force(p1, p2, d, dist2, dist)```

jngrad replied:
that's inelegant for sure, but we need the energy for NPT on the next line, and reducing the use of output parameters is our goal. I don't think there is a performance impact (https://godbolt.org/z/drOVeJ).

KaiSzuttor commented:
could you also please merge the python branch?

mkuron commented on src/core/electrostatics_magnetostatics/icc.cpp:
```diff
@@ -72,11 +72,14 @@ void force_calc_iccp3m(const ParticleRange &particles,
+#ifdef P3M
+  p1->f.f += std::get<1>(forces);
+  p2->f.f += std::get<2>(forces);
+#endif
```
This still modifies the particle. And I think we also add the same force down in forces_inline.hpp.

jngrad replied:
True, it's not a pure function, but it's also not a force kernel. It's only a helper function that is not visible outside this file. The forces here are obtained from `Coulomb::pair_force(p1, p2, d, dist);`, which is indeed called in `forces_inline.hpp`. I might be wrong, but I think this helper function is used by [`iccp3m_iteration()`](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/icc.cpp#L94) to progressively update virtual charges while making sure the resulting particle force doesn't cancel the external electric field. Unfortunately this function isn't documented.

mkuron commented on src/core/electrostatics_magnetostatics/magnetic_non_p3m_methods.cpp:
```diff
@@ -70,51 +54,29 @@ double calc_dipole_dipole_ia(Particle *p1, const Utils::Vector3d &dip1,
+    // Forces
+    auto const ff = ab * dr + cc * dip1 + dd * dip2;
+    p1->f.f += dipole.prefactor * ff;
+    p2->f.f -= dipole.prefactor * ff;
```
Again, this also modifies the particle.

jngrad replied:
Same remark as above.

mkuron reviewed (approved):
Looks ok to me. Note however that I did not match up every single `p->f.f += ...` from before with where it is now though. Those I did look at matched up, so I think this is reasonably safe to merge.

jngrad committed:
Merge branch 'python' into fix-2502-p2

jngrad committed:
Remove prefixes in force/energy kernels

calc_ from bonded/electrostatic/dipole forces
add_ from electrostatic/dipole energies

jngrad committed:
Pass IA parameters by const reference

jngrad committed:
Pass Particles by reference instead of pointers

Only in the force/energy kernels and integration loop.

jngrad committed:
Remove unused functions

jngrad committed:
Return from the formula in force/energy kernels

jngrad committed:
Formatting

jngrad committed:
Remove redundant asserts

RudolfWeeber commented:
>I don't know why NPT requires energies during the calculation of forces. 

The dipolar pressure is proportional to energy / volume.

 

I suggest not to defer this PR for NPT-related reasons.That needs to be refactored, anyway, which is a major project in itself.

The dipolar isotropic NPT-pressure may well be wrong. I don’t think, it’s tested.

RudolfWeeber commented:
bors r+
